src/IBM.Cloud.SDK.Core/Service/IBMService.cs

```diff
@@ -71,272 +56,41 @@ protected string Endpoint
-            {
-                Icp4dUrl = icp4dUrl;
-            }
-
-            if (string.IsNullOrEmpty(IamAccessToken) &&
-                string.IsNullOrEmpty(Icp4dAccessToken) &&
-                string.IsNullOrEmpty(ApiKey) &&
-                (string.IsNullOrEmpty(UserName) || string.IsNullOrEmpty(Password))
-                )
-            {
-                throw new NullReferenceException(string.Format("Either {0}_IAM_ACCESS_TOKEN, {0}_ICP4D_ACCESS_TOKEN, {0}_APIKEY or {0}_USERNAME and {0}_PASSWORD did not exist. Please add credentials with this key in ibm-credentials.env.", ServiceName.ToUpper()));
-            }
-
-            Icp4dConfig icp4dConfig = new Icp4dConfig(
-                url: !string.IsNullOrEmpty(Icp4dUrl) ? Icp4dUrl : null,
-                username: !string.IsNullOrEmpty(UserName) ? UserName : null,
-                password: !string.IsNullOrEmpty(Password) ? Password : null,
-                userManagedAccessToken: !string.IsNullOrEmpty(Icp4dAccessToken) ? Icp4dAccessToken : null
-                );
-
-            IamConfig iamConfig = new IamConfig(
-                apikey: !string.IsNullOrEmpty(ApiKey) ? ApiKey : null,
-                iamUrl: !string.IsNullOrEmpty(IamUrl) ? IamUrl : null,
-                userManagedAccessToken: !string.IsNullOrEmpty(IamAccessToken) ? IamAccessToken : null,
-                iamClientId: !string.IsNullOrEmpty(ClientId) ? ClientId : null,
-                iamClientSecret: !string.IsNullOrEmpty(ClientSecret) ? ClientSecret : null
-                );
-
-            BasicAuthConfig basicAuthConfig = new BasicAuthConfig(
-                username: !string.IsNullOrEmpty(UserName) ? UserName : null,
-                password: !string.IsNullOrEmpty(Password) ? Password : null
-                );
-
-            if (!string.IsNullOrEmpty(ServiceUrl))
-                Endpoint = ServiceUrl;
-
-            if (!string.IsNullOrEmpty(Icp4dAccessToken) && !string.IsNullOrEmpty(Icp4dUrl))
-            {
-                SetAuthenticator(icp4dConfig);
-            }
-
-            if (!string.IsNullOrEmpty(ApiKey))
-            {
-                SetAuthenticator(iamConfig);
-            }
-
-            if (!string.IsNullOrEmpty(UserName) && !string.IsNullOrEmpty(Password))
-            {
-                SetAuthenticator(basicAuthConfig);
-            }
-        }
-
-        protected IBMService(string serviceName, string url)
-        {
-            ServiceName = serviceName;
-            Client = new IBMHttpClient(url, UserName, Password);
-
-            if (!string.IsNullOrEmpty(Endpoint))
-                Endpoint = url;
-            //TODO: verificar como iremos obter de um arquivo json por injeção de dependencia
-            //ApiKey = CredentialUtils.GetApiKey(serviceName);
-            //Endpoint = CredentialUtils.GetApiUrl(serviceName);
+            Client = new IBMHttpClient();
+            authenticator = ConfigBasedAuthenticatorFactory.GetAuthenticator(serviceName);
```

Comment:
I think IBMService should probably not have any references to the external config path at all (i.e. ConfigBasedAuthenticatorFactory).  It should have just a ctor that takes in an IAuthenticator instance.

The .NET generator should generate slightly different code if the "includeDefaultAuthenticator" config property is specified... specifically it should generate an additional ctor that DOESN'T take in an IAuthenticator parameter, but instead calls the ConfigBasedAuthenticatorFactory to instantiate an authenticator from external config info, then pass the authenticator instance into the IBMService ctor.
Let's discuss f2f tomorrow to clarify...

Reply:
OK made the change